services/src/main/java/org/keycloak/services/resources/IdentityBrokerService.java

```diff
@@ -1048,12 +1049,31 @@ private void migrateFederatedIdentityId(BrokeredIdentityContext context, UserMod
 
     private void updateToken(BrokeredIdentityContext context, UserModel federatedUser, FederatedIdentityModel federatedIdentityModel) {
         if (context.getIdpConfig().isStoreToken() && !ObjectUtil.isEqualOrBothNull(context.getToken(), federatedIdentityModel.getToken())) {
-            federatedIdentityModel.setToken(context.getToken());
+            try {
+                // like in OIDCIdentityProvider.exchangeStoredToken()
+                // we shouldn't override the refresh token if it is null in the context and not null in the DB
+                // as for google IDP it will be lost forever
+                if (federatedIdentityModel.getToken() != null) {
```

Comment:
Tests are failing because when the IDP is a SAML IDP, the response can't be read into an OAuth `AccessTokenResponse`.

Changing this line to
```
                if (federatedIdentityModel.getToken() != null && !(context.getIdp() instanceof SAMLIdentityProvider)) {
```
Fixed the test failures for me.